From c3c91f0a1e0337770f76ba58b5690727b7ae6a32 Mon Sep 17 00:00:00 2001
From: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
Date: Tue, 25 Mar 2025 17:06:51 -0700
Subject: [PATCH 06/16] PENDING: net: PCS: Rework and expose qcom_xpcs_destroy
 function

Rework qcom_xpcs_destroy function and include within
header file so that it could be called from stmmac driver
in instances of stmmac driver removal or unbinding
of an ethernet device as We don't need to keep Xpcs to be
active when the ethernet driver/device itself is removed.

Change-Id: I28a44b3e8938253ac3d0d8fab2147d09d3788b65
Signed-off-by: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
(cherry picked from commit 62f6d93a0c71d75dfa5fe1ae1e824cdf3600380f)
---
 drivers/net/pcs/pcs-xpcs-qcom.c | 3 ++-
 include/linux/pcs-xpcs-qcom.h   | 1 +
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/pcs/pcs-xpcs-qcom.c b/drivers/net/pcs/pcs-xpcs-qcom.c
index 4795e0c9a582..4c213f44eac4 100644
--- a/drivers/net/pcs/pcs-xpcs-qcom.c
+++ b/drivers/net/pcs/pcs-xpcs-qcom.c
@@ -711,11 +711,12 @@ struct phylink_pcs *qcom_xpcs_create(struct device_node *np, phy_interface_t int
 EXPORT_SYMBOL_GPL(qcom_xpcs_create);
 
 /* Power off the DWC_xpcs controller */
-void qcom_xpcs_destroy(struct dw_xpcs_qcom *qxpcs)
+void qcom_xpcs_destroy(struct phylink_pcs *pcs)
 {
 	int i, ret;
 	u32 xpcs_id;
 	const struct xpcs_compat *compat;
+	struct dw_xpcs_qcom *qxpcs = phylink_pcs_to_xpcs(pcs);
 
 	xpcs_id = xpcs_get_id(qxpcs);
 	if (xpcs_id == 0xffffffff) {
diff --git a/include/linux/pcs-xpcs-qcom.h b/include/linux/pcs-xpcs-qcom.h
index 9d4cb034f873..2bdb9b65f85c 100644
--- a/include/linux/pcs-xpcs-qcom.h
+++ b/include/linux/pcs-xpcs-qcom.h
@@ -30,4 +30,5 @@ struct phylink_pcs *qcom_xpcs_create(struct device_node *np,
 				      phy_interface_t interface);
 void qcom_xpcs_link_up(struct phylink_pcs *pcs, unsigned int mode,
 		  phy_interface_t interface, int speed, int duplex);
+void qcom_xpcs_destroy(struct phylink_pcs *pcs);
 #endif /* __LINUX_PCS_XPCS_QCOM_H */
